src/irmin/node.ml

```diff
@@ -53,60 +34,84 @@ struct
   type step = P.step
   type metadata = M.t
 
+  type kind = [ `Node | `Contents of M.t ]
+  type entry = { kind : kind; name : P.step; node : K.t }
+
+  let kind_t =
+    let open Type in
+    variant "Tree.kind" (fun node contents -> function
+        | `Node -> node
+        | `Contents m -> contents m)
+    |~ case0 "node" `Node
+    |~ case1 "contents" M.t (fun m -> `Contents m)
+  |> sealv
+
+  let entry_t: entry Type.t =
+    let open Type in
+    record "Tree.entry" (fun kind name node -> { kind ; name ; node } )
+    |+ field "kind" kind_t (function { kind; _ } -> kind)
+    |+ field "name" P.step_t (fun { name ; _ } -> name)
+    |+ field "node" K.t (fun { node ; _ } -> node)
+    |> sealr
+
+  let to_entry (k, v) = match v with
+    | `Node h -> { name = k; kind = `Node; node = h }
+    | `Contents (h, m) -> { name = k; kind = `Contents m; node = h }
+
+  let of_entry n =
+    n.name,
+    match n.kind with
+    | `Node -> `Node (n.node)
+    | `Contents m -> `Contents (n.node, m)
+
   module StepMap =
     Map.Make(struct type t = P.step let compare = Type.compare P.step_t end)
 
   type value = [ `Contents of hash * metadata | `Node of hash ]
 
-  type k =
-    | Map  : value StepMap.t -> k
-    | List: (step * value) list -> k
-    | Both: value StepMap.t * (step * value) list -> k
-
-  type t = { mutable k: k }
-  let of_list l = { k = List l }
-  let of_map m = { k = Map m }
-
-  let list t = match t.k with
-    | List l | Both (_, l) -> l
-    | Map m ->
-      let alist = StepMap.bindings m in
-      t.k <- Both (m, alist);
-      alist
-
-  let map t = match t.k with
-    | Map m | Both (m, _) -> m
-    | List l ->
-      let map =
-        List.fold_left (fun acc (l, x) -> StepMap.add l x acc) StepMap.empty l
-      in
-      t.k <- Both (map, l);
-      map
-
-  let v = of_list
-
-  let find t s =
-    try Some (StepMap.find s (map t))
-    with Not_found -> None
-
-  let empty = of_list []
+  type t = Map of entry StepMap.t
```

Comment:
Why do we need this `Map` constructor ?

Reply:
it's not really needed but improved the type-checking :p I can remove it